Spec: NuGetAudit without nuget.org as a package source (#12918)
NuGet/Home: proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

However, some developers might consider this an unacceptable risk and prefer another way to get a VDB that cannot accidentally be used to get packages.

## Explanation

### Functional explanation

<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
*NuGet.Config* files have a new section `<vulnerabilitySources>` where the URL(s) of any additional sources can be specified, which will be used only to download a VDB, and will not be used for downloading packages.

For example, consider a developer at Contoso using a company-internal source as a single package source, and using nuget.org as a vulnerability source:

```xml
<configuration>
  <packageSources>
    <clear />
    <add key="contoso" value="https://internal.corp.contoso.com/packages/nuget/index.json" >
  </packageSources>
  <vulnerabilitySources>
    <clear />

>     <add key="nuget.org" value="https://api.nuget.org/v3/index.json" >

Reviewer: I have an offline questions about this.

  </vulnerabilitySources>
</configuration>
```

Any URL specified in the `vulnerabilitySources` section must implement the [NuGet Server API V3 protocol](https://learn.microsoft.com/nuget/api/overview).
However, the server will only need to implement the [`VulnerabilityInfo` resource](https://learn.microsoft.com/en-us/nuget/api/vulnerability-info), and can avoid implementing all the other resources that are required to be used as a package source.

### Technical explanation

<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->

> Restore is NuGet's most performance critical code path, so performance should be designed into any implementation.

Reviewer: I'm guessing by this, we're referring to things like no-op? If the projects are up to date, vulnerabilities won't be rechecked. 

Maybe we can just clearly say that vulnerabilities checking is gonna be done in equivalent way regardless of how the source was specified?

Author: I was actually referring to what I wrote underneath, that if the same URL is specified both as an audit source and also a package source, don't load the vulnerability DB twice.

But I updated the spec to suggest to no longer use packageSources as VDB sources when auditSources is used, so this is no longer a concern.

# NuGetAudit without nuget.org as a package source
<!-- Replace `Title` with an appropriate title for your design -->

- Author: [zivkan](https://github.com/zivkan)
- GitHub Issue: [12698](https://github.com/NuGet/Home/issues/12698)

## Summary

<!-- One-paragraph description of the proposal. -->
This proposal adds a new `<auditSource>` to *NuGet.Config* files, allowing *NuGetAudit* to obtain known vulnerabilities without adding nuget.org as a package source.

## Motivation

<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
[NuGetAudit](https://learn.microsoft.com/nuget/concepts/auditing-packages) is a feature where NuGet will report during restore which packages with known vulnerabilities are being used by projects, [proposed in late 2022](../2022/vulnerabilities-in-restore.md), and finally released with .NET 8 in 2023.

> The first version of NuGetAudit used package sources as the data source for known vulnerabilities, which will be referred to as a Vulnerability Database (VDB) for the remainder of this document.

Reviewer 2: FYI a better acronym for the database and related items might be:

OSV - Open source vulnerabilities -> https://osv.dev/

OSVDB - Open source vulnerability database -> https://github.com/advisories

Author: There's nothing in NuGet that requires the vulnerability information to be limited to open source libraries. If closed source package providers (I'm thinking the ASP.NET & WPF control vendors) want to provide advisories on their closed source packages, it'll work just fine.  I also don't want to want to give a false impression that NuGet's using one of these formats. We created our own, minimal, file format in order to maximize performance.

While nuget.org provides a VDB, there are several reasons why customers may not be using nuget.org as a package source.
This is especially true following the 2021 blog post titled "Dependency confusion" which significantly increased knowledge of multi-source package substitution supply chain risks.

While the design of NuGetAudit attempts to make it easy for upstreaming package sources to also upstream the `VulnerabilityInfo` resource, the NuGet team and customers are reliant of the feed implementor to implement the change.

> However, there are also cases where customers are not using upstreaming feeds, but instead curate approved 3rd party packages and manually upload them to an internal feed.

Reviewer 2: There are many cases. This specific scenario is common in Microsoft, but not the outside world per-say.

Author: I intended this to be a non-exhaustive example. I'll try to make that clearer.

What upstreaming means is a feed that is not solely a repository for packages that were pushed to the feed, but can also automatically obtain packages from other feeds.

It's possible for customers to use [Package Source Mapping](https://learn.microsoft.com/nuget/consume-packages/package-source-mapping) to add nuget.org as a package source, but configured in a way that NuGet will never restore packages from this package source.
However, some developers might consider this an unacceptable risk and prefer another way to get a VDB that cannot accidentally be used to get packages.

## Explanation

### Functional explanation

<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->

> *NuGet.Config* files have a new section `<auditSources>` where the URL(s) of any additional sources can be specified, which will be used only to download a VDB, and will not be used for downloading packages.

Reviewer 2: Love it.

> *NuGet.Config* files have a new section `<auditSources>` where the URL(s) of any additional sources can be specified, which will be used only to download a VDB, and will not be used for downloading packages.

Reviewer 3: > *NuGet.Config* files have a new section `<auditSources>` where the URL(s) of any additional sources can be specified

Would it be possible to download VDB from multiple audit sources? In other words, can customers configure multiple audit sources in the nuget.config file?

Author: Yes. I hoped that the example below that uses `<clear />` and `<add` made that sufficiently clear, despite not having an example with 2 audit sources. But I can add a sentence to explicitly state it.

edit: I feel like the language is already clear that there can be several sources. Firstly, the plural word sources is used every time, never the singular source. Secondly, "URL(s)" implies that there can be one or more. The sentence immediately after the one this comment is attached to says "at least one source".

### Technical explanation

<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->

Not using `packageSources` for VDB when there's at least one source in `auditSources` is a performance optimization.
Since we expect customers to use `auditSources` when their `packageSources` don't already provide a VDB, having restore spend time making HTTP requests to the package sources is unlikely to provide any benefit.
While downloading the VDB from audit sources is likely to take more time than verifying that package sources don't have a vulnerability resource, if any package source is slow, downloading the VDB from the audit source might complete first.
Additionally, in the case where NuGet doesn't otherwise need to communicate with any package source, checking the package sources for a VDB might trigger a credential provider.

## Drawbacks

<!-- Why should we not do this? -->
Customers who appear functionally network isolated from nuget.org (for example, by blocking nuget.org at the firewall) will still not be able to use this feature.

> ## Rationale and alternatives

Reviewer 2: What about .csproj based definition for modern .net scenarios?

Author: I'll add a section. Basically, I'm unconvinced it adds anything useful. I know there are some "MSBuild everything" individuals, but I don't see anything compelling about being able to add additional audit sources via msbuild.

<!-- Why is this the best design compared to other designs? -->
<!-- What other designs have been considered and why weren't they chosen? -->
<!-- What is the impact of not doing this? -->

> ### Wait for other servers to implement the VulnerabilityInfo resource

Reviewer 2: Great explanations!

While the NuGet team may be able to influence other teams and encourage them to work on implementing support for the `VulnerabilityInfo` resource, ultimately the teams that implement those NuGet servers have their own backlogs and priorities.
Therefore, the NuGet team doesn't have agency to ensure that it gets implemented, and customers who want to use *NuGetAudit* may not have enough influence to get the implementation prioritized in a timely manor.

> Additionally, not all customers use a server that supports upstreaming, or have configured their server to use nuget.org as an upstream source.

Reviewer 3: If a customer is using an Azure DevOps feed with `nuget.org` as one of the [upstream sources](https://learn.microsoft.com/en-us/azure/devops/artifacts/concepts/upstream-sources?view=azure-devops), does NuGet audit functionality be able to download VDB from nuget.org?

Author: upstream sources isn't a concept that NuGet (client) has, so servers that support "upstream sources" have to make it completely opaque to NuGet. This section of the spec is specifically about the point that servers might not have implemented upstreaming vulnerability info from their upstream source(s).

The motivation section at the top of the document also mentioned upstream sources.

For example ensuring it's patched following every Patch Tuesday, and probably maintaining docker images to make it easy to use.

Unlike this spec's proposal, this vulnerability-only upsourcing server idea could in theory enable customers in companies that block nuget.org at the firewall to get always up-to-date VDB.
But in such a restrictive environment, I expect it's even less likely that their security would allow this web app to be installed than other companies that simply have a policy of "don't use nuget.org as a package source".

The NuGet team already has a [NuGet server implementation](https://github.com/NuGet/NuGet.Server) that can be self hosted.
But the V3 NuGet Server API wasn't implemented.
While I find the idea of a reference implementation of NuGet's server API interesting (imagine ASP.NET Core middleware where customers implement a backend interface, and getting new NuGet features is as simple as upgrading the middleware package version), we should consider [NuGet.Server](https://github.com/NuGet/NuGet.Server)'s lack of modernization as a signal that it's likely an unreasonable cost.

> ### Read vulnerability database from local files

Reviewer 2: Can I read or override with any OSV / OSVDB .json/.zip from a webserver too if a server doesn't implement vulnerability functionality? i.e. 

https://google.github.io/osv.dev/data/#current-data-sources | https://github.com/github/advisory-database#sources

https://storage.googleapis.com/osv-vulnerabilities/index.html?prefix=NuGet/ (see all.zip)

Author: Currently, no. The original NuGetAudit spec proposed a custom file format, and I don't remember anybody commenting asking for other formats to be readable.

   Should we just add it as a default in new user-profile nuget.config files, and just accept that existing developers won't get it by default?

2. Stop using package source as audit source?

   On one hand, using package sources as a defacto audit source makes it easier for customers to use the feature when they knowingly or unknowingly use a package source with a VDB.
   On the other hand, overloading `<packageSources>` with multiple meanings is confusing and customers probably won't understand why sometimes they get vulnerable package information and sometimes they don't.
   If we add nuget.org as a default audit source, and don't use packageSources when auditSources has a value, then using packageSources to get a VDB when auditSources is empty will be particularly confusing/unintuitive to customers.

   Another challenge is that if we want to change the behavior it' will be a breaking change, which is typically discouraged.
   However, given that nuget.org is the only know source with a VDB, if we make nuget.org a default audit source, then it will effectively be a non-breaking change.

> 3. Tooling

Reviewer 2: I would not prioritize this for now. This should be a pretty advanced scenario not requiring tooling.

Upstreaming isn't a concept that exists in NuGet, so servers have to present to clients as if all data (packages, search results, vulnerability information) originates from itself.

However, there are also cases where customers are not using upstreaming feeds, but instead curate approved 3rd party packages and manually upload them to an internal feed.
There are also other scenarios where customers do not use nuget.org (directly) as a package source.

It's possible for customers to use [Package Source Mapping](https://learn.microsoft.com/nuget/consume-packages/package-source-mapping) to add nuget.org as a package source, but configured in a way that NuGet will never restore packages from this package source.
However, some developers might consider this an unacceptable risk and prefer another way to get a VDB that cannot accidentally be used to get packages.

## Explanation

### Functional explanation

<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->

> *NuGet.Config* files have a new section `<auditSources>` where the URL(s) of any audit sources can be specified, which will be used only to download a VDB, and will not be used for downloading packages.

Reviewer 4: Looks like this ie more of `alternateAuditSources` based on what's described here?

Author: No, but I think your other comment explains why you came away with this conclusion: https://github.com/NuGet/Home/pull/12918#discussion_r1372141490

I'll continue the discussion there, as it has more context about the interpretation of the feature.

Like `<packageSources>`, `<auditSources>` can contain zero, one, or more sources, and supports `<clear />`.

> When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.

Reviewer 4: My interpretation without documentation for this keyword would be that all `packageSources `are `auditSources `by default and any source mentioned as `auditSources` would be additional audit source unless `</clear>` is used which will obviously clear all packageSources as audit sources

Author: That's an interesting interpretation, but not what I intended.

As you know, nuget accumulates settings from multiple nuget.config files. Therefore, `<clear />` within an `<auditSources>` means "if a lower priority nuget.config file has already defined any audit source, remove it". This is important in CI scenarios where a "stateful" agent is used, since another build on the same agent might have modified the user-profile nuget.config file.

Once NuGet has generated the effective settings from all nuget.config files, NuGet now has a list of packageSources and potentially a list of audit sources. If audit sources list length is zero, then package sources are used as audit sources. Otherwise, only audit sources are used to download a vulnerabilities database.

Reviewer 2: another perspective here that i'm seeing from anand's feedback is that yes we absolutely imply a package source is an audit source, but the specific use-case we are trying to address here is when you don't have a package source that you want to use for auditing data. 

so in all, if you specified nuget.org as a package source, you wouldn't need to imply it as an audit source as it would be implicit.

but if you say run only on an internal source, you may absolutely want to use the nuget.org audit source. this is what i meant earlier by advanced scenario. 

the core scenario would be that people are using package sources that have auditing capabilities enabled. but the adv scenario is they override it using this wonderful spec of `<auditSources>`.

Author: I don't understand what's actionable on my side. Are you (both?) proposing that I remove the line that this comment thread is linked to (don't use package sources as audit sources, when there's at least one audit source explicitly defined). Are you saying you want the spec changed so that `auditSources` is **always** additive to the list of `packageSources`?

My position is that it was a mistake to use package sources as audit sources in the first place. Overloading a single thing for multiple reasons just generally complicates things. I don't think that having customers duplicate feed URLs (same URL in both `packageSources` and `nugetSources` sections) is a problem. However, completely stopping using package sources as audit sources would be a breaking change from NuGet 6.8. Hence why this spec proposes using packageSources for NuGetAudit only when zero `auditSources` are defined.

While it can be argued that using package sources for NuGetAudit makes is "easier" for customers to use, it has the unintended consequence that authenticated feeds that don't provide vulnerability info may cause NuGet to prompt interactively for credentials when there's no benefit (no packages needed from the feed). Perhaps some people will think this is a small edge case that doesn't matter, but I don't agree, and the point is that it's not a "zero downside" design choice, or even "obviously better, noone will disagree" design choice.

Reviewer 2: The actionable part here I believe is that we should ensure that any defined packageSource that supports VDB is implicitly added to the auditSources (doesn't have to be written to file).

The customer should be able to include a reputable package source in their packageSources and for features to "just work". When a customer has an advanced use-case, they use the auditSources overriding behavior here.

For example, consider a developer at Contoso using a company-internal source as a single package source, and using nuget.org as an audit source:

```xml
<configuration>
  <packageSources>
    <clear />
    <add key="contoso" value="https://internal.corp.contoso.com/packages/nuget/index.json" >
  </packageSources>
  <auditSources>
    <clear />
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" >
  </auditSources>
</configuration>
```

Any URL specified in the `auditSources` section must implement the [NuGet Server API V3 protocol](https://learn.microsoft.com/nuget/api/overview).
However, the server will only need to implement the [`VulnerabilityInfo` resource](https://learn.microsoft.com/en-us/nuget/api/vulnerability-info) (and service index), and can avoid implementing all the other resources that are required to be used as a package source.

> 

Reviewer 4: [Non MVP, could wait for customer feedback] What's the format for popular audit sources? Could this support some generic format as well like say a csv file?

Author: It's officially documented on our docs site: https://learn.microsoft.com/en-us/nuget/api/vulnerability-info

It was proposed in the original NuGetAudit feature: https://github.com/NuGet/Home/blob/dev/proposed/2022/vulnerabilities-in-restore.md#dedicated-vulnerability-information-resource

I'm resolving this conversation as out of scope, since this spec is about a specific NuGetAudit feature, not a broad NuGetAudit discussion.

> Could this support some generic format as well like say a csv file?

This is already addressed at the end of the spec.

### Technical explanation

<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->

> Not using `packageSources` for VDB when there's at least one source in `auditSources` is a performance optimization.

Reviewer: I'm not sure this performance optimization adds much, but I do think this has a notable UX experience. 

I'm not expecting many sources to build vulnerability databases if we add this, so the impact is small, but this is a UX change. It may require you to double specify sources.

Author: It's definitely not a significant perf improvement. 

Just to make sure I understand what you mean by "double specify sources", are you thinking of a scenario where a customer adds nuget.org as an audit source, but they also want vulnerability info from one of their package sources, so now they're going to have to specify that package source both in package sources and audit sources?  If you mean something different, then it's not intuitive to me (and frankly this interpretation of your comment feels like a guess)

On the other hand, consider a scenario where a project removes a PackageReference, or switches between branches (both of which have previously been restored, so global packages folder is populated). There's a good chance that NuGet won't need to query any package sources during the graph resolution. If NuGetAudit contacts package sources for the vulnerability info, then that might trigger interactive credential acquisition. So, the ability to not use all package sources as audit sources can also be considered a UX improvement.

But yeah, it's a UX change. As mentioned during the meeting, I regret the initial NuGetAudit implementation. I don't like package sources being overloaded for multiple purposes. I think it would be less confusing to customers if audit sources were only explicitly defined. If I could have gone back in time, I would have made this the first implementation, and if we got feedback then change it to what we currently have. But there are many examples of breaking changes making CX improvements difficult or impossible 😢

Reviewer: > Just to make sure I understand what you mean by "double specify sources", are you thinking of a scenario where a customer adds nuget.org as an audit source, but they also want vulnerability info from one of their package sources, so now they're going to have to specify that package source both in package sources and audit sources

Yes, you'd basically need to specify nuget.org in both sources & audit sources.

Now I see that your spec says, when there's at least one. 
Not really sure how I feel about the whole thing tbh, it's a bit confusing, it doesn't affect people who have nuget.org in their sources already, so that concern is addressed.

Since we expect customers to use `auditSources` when their `packageSources` don't already provide a VDB, having restore spend time making HTTP requests to the package sources is unlikely to provide any benefit.
While downloading the VDB from audit sources is likely to take more time than verifying that package sources don't have a vulnerability resource, if any package source is slow, downloading the VDB from the audit source might complete first.

> Additionally, in the case where NuGet doesn't otherwise need to communicate with any package source, checking the package sources for a VDB might trigger a credential provider.

Reviewer: What about auth?
Do audit source names need to match package source names? 
Do credentials apply to audit sources as well?

The reasons customers may do this do not preclude them from using nuget.org as an audit source, so allowing these customers an easy way to use *NuGetAudit* benefits them.

Therefore, in order to provide customers with the ability to use NuGetAudit most easily, it's more effective to implement a new feature that does not depend on other teams to implement features.

### NuGet Team supplied upsourcing server

The NuGet team could provide a web application that looks like a NuGet package source that never contains any packages (and doesn't support push), and the `VulnerabilityInfo` resource effectively proxies nuget.org's VDB.
However, getting permission to install a new web app is likely difficult in most corporate environments, so wouldn't be practical.

Besides the high up-front cost of implementation, it would also have ongoing maintenance costs to the NuGet team.
For example ensuring it's patched following every Patch Tuesday, and probably maintaining docker images to make it easy to use.

Unlike this spec's proposal, this vulnerability-only upsourcing server idea could in theory enable customers in companies that block nuget.org at the firewall to get always up-to-date VDB.

> But in such a restrictive environment, I expect it's even less likely that their security would allow this web app to be installed than other companies that simply have a policy of "don't use nuget.org as a package source".

Reviewer: Another idea is to a static source that just points to nuget.org. 

That doesn't help the people blocking with firewalls, but it helps the people that just don't want to use nuget.org

### Read vulnerability database from local files

In order to enable customers working in offline environments (or just nuget.org is blocked), the most realistic option is likely to allow NuGet to read the VDB from a file on disk (including a network share).
In these disconnected (from nuget.org) environments, it will be up to the customer to find an approved way with their security compliance to get nuget.org's VDB onto their computer.
In non-disconnected scenarios, we could provide a tool, such as `dotnet tool install -g NuGet.VulnerabilityTool ; NuGet.VulnerabilityTool download -o path/to/destination/`.

However, this doesn't provide any benefit over this spec's proposed *NuGet.Config* `<auditSource>`, coupled with the future possibility's [auditSources with local files](#allow-auditsources-to-point-to-a-filesystem-directory) idea.
It would be worse because for customers who do not have nuget.org blocked at the firewall, it would require a manual action to update the VDB, whereas an audit source would allow NuGet to automatically download and cache the VDB.

> ### Specify audit sources via MSBuild

Reviewer: You can just add a property "UseDefaultAuditSource`, and that just adds nuget.org when specified.

No need to have capability to add other audit sources.

Author: I removed the question about whether or not we should enable nuget.org as an implicit default, so I no longer think this suggestion makes sense. So, that's why I'm resolving this conversation without updating the document.

However, we have not received much feedback that this is desirable for package sources, so it's unlikely that it's important for audit sources.

## Prior Art

<!-- What prior art, both good and bad are related to this proposal? -->
<!-- Do other features exist in other ecosystems and what experience have their community had? -->
<!-- What lessons from other communities can we learn from? -->
<!-- Are there any resources that are relevant to this proposal? -->
Although [`npm audit`'s help docs](https://docs.npmjs.com/cli/v10/commands/npm-audit?v=true) don't mention it, it appears to have a `--registry={url}` argument, as can be seen on [the Azure DevOps docs on npm audit](https://learn.microsoft.com/azure/devops/artifacts/npm/npm-audit?view=azure-devops&tabs=yaml).

Pypi's `pip` has [an audit command](https://pypi.org/project/pip-audit/), and `--index-url` and `--extra-index-url` appear to enable using custom audit sources.

Rust's cargo audit command has [configuration for the advisory DB URL](https://docs.rs/cargo-audit/latest/cargo_audit/config/struct.DatabaseConfig.html)

> ## Unresolved Questions

Reviewer 5: Does it make sense to require HTTPS for vulnerablility info?  For the kind of information stored I can see how it would make sense to fully trust the source but we've received negative feedback around requiring HTTPS for package sources in the past.

Reviewer 2: would this be covered by our .net 8 warnings and .net 9 errors?

Reviewer: I think that was added in the newest version of the spec, audit sources will be treated the same as package sources.

<!-- What related issues would you consider out of scope for this proposal but can be addressed in the future? -->

1. default value?

   Should we add nuget.org as a default value, so that customers who remove nuget.org as a package source are more easily able to use NuGetAudit?
   Would using nuget.org as a default audit source cause undue concern from customers who don't want to use nuget.org as a package source, and the new tooling version starts making network requests to nuget.org?

   If we want to make nuget.org a default/automatic audit source, how?
   NuGet has tried before to use a tracking file to automatically add nuget's v3 URL as a package source to user-profile nuget.config files that don't have any package sources.
   However, this caused a security issue when customers wanted to remove nuget.org as a default package source, and this code incorrectly re-added it once (the customer could successfully remove nuget.org a second time).
   When we removed this auto-update to mitigate the security concern, it exposed a bug where multiple other apps created user-profile nuget.config files without any package sources, and many customers assumed it was a bug with NuGet.

>    Should we hard-code nuget.org's default URL in NuGet.Configuration, and require customers to use `<clear/>` if they don't want any audit source?

Reviewer 6: This feels weird. As a new feature, don't force adoption, and we can consider later how we might funnel customers into it?

Author: On the other hand, learning from our team's experience of trying to migrate from NuGet's V2 URL to the V3 URL, and how difficult it is to get existing developers learn about and adopt changes to the default nuget.config, I was looking for other ideas.

Additionally, while I'm not an expert in any other package system, I believe that for npm, pypi, cargo, and probably others, you don't need a configuration file that states the default package repository, or audit service url. My understanding it's hardcoded into the package manager, and if you don't want to use the official package source, then you need to explicitly configure that.

NuGet on the other hand is almost (but not quite) the opposite. nuget.org is only used if explicitly stated in config files, which NuGet will create when no user-profile config file exists, but this means that every developer has a config file, even when they're happy to use defaults, which means it's next to impossible for us to actually change defaults, since it requires changes to the config files.

And that's exactly the problem we had with the V2 to V3 URL migration. NuGet can't tell the difference between "customer who is using the default, so we can/should change" vs "customer explicitly wants to use this setting, which happens to be the old default".

   Should we just add it as a default in new user-profile nuget.config files, and just accept that existing developers won't get it by default?

> 2. Stop using package source as audit source?

Reviewer 6: Haven't you already answered this implicitly? The inverse of what you say in the quote below, I believe, is that *not* enabling any `auditSource`, means the current behavior remains. That is, continue looking for `VulnerabilityInfo` in the `packageSource` and if it's there, *do use it* from the packageSource's resource. In practical terms, that means `nuget.org` will be the VDB. No need to stop this from what I can tell.

> When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.

Author: @jeffkl wrote something related in his PR review comment, which GitHub infuriatingly doesn't make a "conversation", making it much harder to respond to and discuss (have a conversation about).

> Haven't you already answered this implicitly?

No. I wish I knew what it was about my writing that I could improve to reduce misunderstandings. Having said that, when I feel like I'm under time pressure, I don't read other peoples specs carefully enough to understand subtle differences either 😕 

What I intended in this section was specifically to make it so `packageSources` are **NEVER** used as audit sources. The earlier proposal to use packageSources when no auditSources are defined is solely for backwards compatibility reasons. If I could go back in time a few months to before NuGetAudit V1 was merged, I would implement it requiring auditSources as the only way to define which source provides audit info, no fallback to packageSources.

Backwards compatibility is important, but as NuGetAudit is a brand new feature, there's next to zero adoption. In fact, NuGetAudit isn't even GA yet, but it's too late to make any changes to what will ship as GA now. So whatever this spec proposes/implements will be changes to GA behaviour. But, being a new feature, the earlier we make "breaking" changes, the fewer customers will be impacted. If we wait until we customers start using the feature, and several customers give us feedback saying they think some other behaviour would be better, then by definition there will be a lot more people impacted by the change.  So, I thought there'd be opportunity to make a potential breaking change while adoption is low, so most people would never notice.

Reviewer 6: Watched the meeting recording and it seems like decision were made. I'll re-review the latest iteration.

   On one hand, using package sources as a defacto audit source makes it easier for customers to use the feature when they knowingly or unknowingly use a package source with a VDB.
   On the other hand, overloading `<packageSources>` with multiple meanings is confusing and customers probably won't understand why sometimes they get vulnerable package information and sometimes they don't.
   If we add nuget.org as a default audit source, and don't use packageSources when auditSources has a value, then using packageSources to get a VDB when auditSources is empty will be particularly confusing/unintuitive to customers.

>    Another challenge is that if we want to change the behavior it' will be a breaking change, which is typically discouraged.

Reviewer 6: ```suggestion
   Another challenge is that if we want to change the behavior it will be a breaking change, which is typically discouraged.
```

   However, given that nuget.org is the only know source with a VDB, if we make nuget.org a default audit source, then it will effectively be a non-breaking change.

3. Tooling

   How important is it to add CLI commands to manage audit sources?

>    How important is it to add a GUI in Visual Studio to manage audit sources?

Reviewer 6: Another spec could cover this in a future iteration, IMO. Certainly, some indicator whether any audit source is available might be interesting. Does a UI showing 0 vulnerabilities mean: 
"there's no vulnerabilities or there's no audit source active in my view?"

Author: I originally had a warning for "NuGetAudit is enabled, but no sources provided vulnerability info" in NuGet 6.7 (maybe also 6.8? I can't remember). But it was ripped out because rather than having a specific MSBuild property to enable this warning, it was enabled by explicitly setting `<NuGetAudit>true</NuGetAudit>` in the project. When `<NuGetAudit>` was not defined in a project, audit as a whole was still enabled, just the "no vulnerability data available" warning was disabled. This made `<NuGetAudit>` a tri-state, not a boolean on/off.

Anyway, when I was discussing this (I think with @JonDouglas, but I forgot with who exactly), whoever I was talking to adviced me to just remove the warning, and wait for customers to give us feedback that they actually want it. So, if someone thinks it's a good idea, I recommend creating a new feature request

Since this spec is specifically about being able to set audit sources that are not package sources, I'm resolving this conversation as out of scope.

> 4. Restore summary

Reviewer: my vote is yes.

   When you restore at normal verbosity, NuGet outputs a list of package sources that were used for restore. Should we add audit sources as well?

>    If so, presumably we should only output it if at least 1 project in the restore has NuGetAudit enabled, as outputting audit sources when NuGetAudit is disabled for all projects might confuse customers.

Reviewer 6: Isn't this just a flat list that isn't correlated to projects? How will it indicate which projects were analyzed and which weren't?

Author: I think there's a misunderstanding. Since this is one of the "additional questions" that I'm going to delete before the spec is merged, I didn't bother putting effort into an example. But what I'm talking about is this:

```text
         NuGet Config files used:
             C:\Users\zivkan\AppData\Roaming\NuGet\NuGet.Config
             C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.FallbackLocation.config
             C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.Offline.config

         Feeds used:
             https://api.nuget.org/v3/index.json
             C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\
             C:\Program Files\dotnet\library-packs
````

This question is about adding an additional bit:

```text
         Audit sources used:
             https://api.nuget.org/v3/index.json
```

> How will it indicate which projects were analyzed and which weren't?

This spec is intended to be specific about being able to audit sources which are not package sources. Other generic NuGetAudit features were intentionally out of scope.

Your suggestion overlaps greatly with something already proposed in the original NuGetAudit (whole of feature) spec: https://github.com/NuGet/Home/blob/dev/proposed/2022/vulnerabilities-in-restore.md#command-line-restore-output

We also have a (private) tracking issue: https://github.com/NuGet/Client.Engineering/issues/2340  Why is it a private tracking issue? I don't know. Seems like an oversight now. Customers should be able to upvote the idea to give us feedback they think it's more or less important than other features.

Reviewer 6: > if at least 1 project in the restore has NuGetAudit enabled,

When we output those "Feeds used", those are applying to a solution.
Similarly, the "NuGet Config files used:" are solution-level concepts.
Can you explain how a customer will know that "Audit sources used:" is relevant to a project as your comment suggests?

Author: Since audit sources can't be defined by MSBuild, then whatever set of audit sources are listed in the nuget.config apply to all projects. It's just that projects that disable the entire NuGetAudit feature won't use them, but they're still the "defined" audit sources. This is not conceptually different to package sources ("feeds used") when a project doesn't use any packages (either directly or transitively).

I believe your comment/question is more about "how do customers get feedback about which projects have NuGetAudit enabled or disabled". This feature spec (aimed at using nuget.org vulnerability database without also using it as a package source) is orthogonal to that question. I can understand why you're bringing it up, since this is discussing showing something in the restore summary that wasn't being shown earlier. But I feel like this feature doesn't make the "do individual projects enable/disable audit" any better or worse.

The original NuGetAudit spec from last year included generalized restore summary (including a proposed output format), although it was moved to the "future possibilies" section. There's also [this tracking issue](https://github.com/NuGet/Client.Engineering/issues/2340) that I linked to in a few other "conversations" in this PR.

Reviewer 6: > I believe your comment/question is more about "how do customers get feedback about which projects have NuGetAudit enabled or disabled"

Yes, exactly. I chose "indicate" and you chose "get feedback", but we're thinking the same.
Issue 2340 seems to address my concern.

## Future Possibilities

<!-- What future possibilities can you think of that this proposal would help with? -->

> ### Allow auditSources to point to a filesystem directory

Reviewer 6: I was thinking about offline scenarios as I read the spec, and was glad to see you mentioned it. However, I don't see this as a proper "Alternative", which I define as, "we decided this implementation isn't as good as the one proposed". 

Instead, this seems like a phase/iteration we should expect for this work. For the reasons you mentioned, I think supporting offline scenarios are important. The implementation doesn't need to be defined in this spec, but I'd like to see something concrete that it's a "real thing we will dig into in a (not too distant) future iteration" and not "an alternative we are casting out".

Author: This block you've commented on is under the "future possibilities" section, not "rationale and alternatives".

Reviewer 6: You mentioned it in both sections. You even provided an in-document hyperlink to the Future Possibilities section from the [Rationale and alternatives](https://github.com/NuGet/Home/blob/zivkan-vulnerabilitySource-spec/proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md#rationale-and-alternatives) section. I feel it should be in "future" and not "alternatives".

https://github.com/NuGet/Home/blob/zivkan-vulnerabilitySource-spec/proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md#read-vulnerability-database-from-local-files

Author: I also have it under alternate designs, because I feel like implementing this first will give customers a workaround the "don't want to use nuget.org as a package source" scenario, and therefore `auditSources` would be unnecessary, and could save us from implementing the more complex design. I have it under both sections, because as an alternate design, it has UX issues, and I wanted to explain why it wasn't the chosen design.

Reviewer 6: I watched the recording going through this spec, and reread this section. I think the "alternative" is that we're not shipping a dotnet tool that reads VDB files, and the "future" section is that we *will* support local audit sources? I hope your definition of the future section is such that it means an iteration or two from now, it'll be supported.

Reviewer 6: Is there an issue for the local sources that can be linked here?
I would upvote it if I weren't on the team. Hopefully those in offline environments would see it and upvote it if interested in the feature.

Author: All the issues I create around NuGetAudit all have NuGetAudit in the title, making them [easy to search for](https://github.com/NuGet/Home/issues?q=is%3Aissue+NuGetAudit+is%3Aopen+).

I hadn't created one before, so I created one now:
* https://github.com/NuGet/Home/issues/12975

I'm not confident that many people will actually want this, so I marked it as in the icebox. We'll see if it gets enough upvotes before the bot closes it.

<!-- One-paragraph description of the proposal. -->
This proposal adds a new `<auditSource>` to *NuGet.Config* files, allowing *NuGetAudit* to obtain known vulnerabilities without adding nuget.org as a package source.

## Motivation

<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
[NuGetAudit](https://learn.microsoft.com/nuget/concepts/auditing-packages) is a feature where NuGet will report during restore which packages with known vulnerabilities are being used by projects, [proposed in late 2022](../2022/vulnerabilities-in-restore.md), and finally released with .NET 8 in 2023.
The first version of NuGetAudit used package sources as the data source for known vulnerabilities, which will be referred to as a Vulnerability Database (VDB) for the remainder of this document.
While nuget.org provides a VDB, there are several reasons why customers may not be using nuget.org as a package source.
This is especially true following the 2021 [blog post titled "Dependency confusion"](https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610) which significantly increased knowledge of multi-source package substitution supply chain risks.

> Someone at Microsoft published [a whitepaper titled "3 ways to mitigate risk when using private package feeds"](https://azure.microsoft.com/mediahandler/files/resourcefiles/3-ways-to-mitigate-risk-using-private-package-feeds/3%20Ways%20to%20Mitigate%20Risk%20When%20Using%20Private%20Package%20Feeds%20-%20v1.0.pdf), so from customers points of view, it appears that Microsoft's official advice is to not use nuget.org directly, but instead use a single package source that uses nuget.org as an upstream source.

Reviewer: I think this contain too strong wording. 
I think we should rephrase to call out PSM as a mitigation given that the paper is out of date. 
That paper is about solutions at that point in time, imo it's not necessarily an official advice.

There are also other scenarios where customers do not use nuget.org (directly) as a package source.

It's possible for customers to use [Package Source Mapping](https://learn.microsoft.com/nuget/consume-packages/package-source-mapping) to add nuget.org as a package source, but configured in a way that NuGet will never restore packages from this package source.
However, some developers might consider this an unacceptable risk and prefer another way to get a VDB that cannot accidentally be used to get packages.

## Explanation

### Functional explanation

<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->

#### auditSources in nuget.config

*NuGet.Config* files have a new section `<auditSources>` where the URL(s) of any audit sources can be specified, which will be used only to download a VDB, and will not be used for downloading packages.
Like `<packageSources>`, `<auditSources>` can contain zero, one, or more sources, and supports `<clear />`.

> When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.

Reviewer 6: > When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.

This feels elegant to me. Instead of the previous text that meant "falling back to `NuGet.Configuration` for a default audit source", I like that it is either implicit that package source's vulnerability resource is used, unless `<auditSources>` is provided.

Author: What you're referring to was only in an "unanswered questions" section from an earlier version of the spec. That proposal was part of changing NuGetAudit to never use package sources as audit sources, and therefore we'd need another way to set up default audit sources in order to allow NuGetAudit to work by default.

It's precisely how other package ecosystems work (the tool has a hardcoded default, for both package sources, and their audit URL). NuGet supports `<clear/>` on all these "list data types", which would be a way for customers to remove the default, if we had implemented it this way instead.

My gut feeling is that using package sources as audit sources will cause more confusion than benefit, in the long term, and I don't believe that customers needing to duplicate a URL in two sections of the config file is a serious issue. But unfortunately we took an easy option with NuGetAudit V1 and now it would be a breaking change to change the behaviour to what I consider a better option.

Author: I'm resolving this conversation as nothing actionable for me to modify in the design spec. If you don't agree, please clarify your feedback.

For example, consider a developer at Contoso using a company-internal source as a single package source, and using nuget.org as an audit source:

```xml
<configuration>
  <packageSources>
    <clear />

>     <add key="contoso" value="https://internal.corp.contoso.com/packages/nuget/index.json" >

Reviewer 6: What this means is that `contoso` will *not* have its vulnerability resource checked, right?

If the `<auditSources>` tab were missing or did not contain a `<clear/>`, then both `contoso` and `nuget.org` would be checked for vulnerability resources?

Author: > What this means is that contoso will not have its vulnerability resource checked, right?

correct

> If the `<auditSources>` tab were missing or did not contain a <clear/>, then both contoso and nuget.org would be checked for vulnerability resources?

No. This is [the same interpretation as what Anand had](https://github.com/NuGet/Home/pull/12918#discussion_r1372141490).

The `<clear />` only means to remove any `auditSources` defined in nuget.org files, which is the same semantics as what `packageSources`, `packageSourceCredentials`, `packageSourceMapping`, and anything else in nuget.config that supports `<clear/>`.

I don't know NuGet.Configuration's implementation, so I don't know how easy or difficult it would be to have `auditSources` effectively auto-import all `packageSources`, and therefore use `<clear/>` to not only clear `auditSources` defined in different nuget.config files, but also clear `packageSources` that auto-flowed. But then we get into more complex scenarios like what if `c:\src\nuget.config` defines PkgSrc1 as a package source, and nuget.org as an audit source (without a clear), then `c:\src\sln1\nuget.config` sets up only `packageSources` with a clear and PkgSrc2, then a URL, but does not have an audit sources section.  What would the resulting audit sources be? PkgSrc1 and nuget.org because that's what was inherited from the parent directory? PkgSrc2 and nuget.org, because the "accumulated" package sources is PkgSrc2 only?

Also, would there be a desire to "clear inherited audit sources defined in a parent nuget.config, but I still want to include package sources as audit sources, in addition to adding an additional audit source?

This is why I really don't like using `packageSources` as a fallback for `auditSources`. It adds too much complexity and therefore confusion. I don't think defining the same URL twice, once for package sources and once for audit sources is a serious enough "problem" to warrant implicitly using the package sources, when customers do want to use the same source for both.

However, because we don't like to introduce breaking changes, this is the only reason I proposed continuing to use `packageSources` as a fallback when the `auditSources` list is empty. It would be less confusing for everyone is NuGetAudit never used `packageSources` to get audit information. But, it would be a breaking change.

This spec does not provide any way to stop using package sources as an audit source, without also explicitly defining at least 1 audit source. Because what that would mean is to disable NuGetAudit, which they can do via `<NuGetAudit>false</NuGetAudit>` in their project file or Directory.Build.props.

Author: I'm resolving this conversation as a duplicate of the one I linked. Please comment there to help me drive this spec to something that's sufficiently agreed upon so it can be merged.

    <clear />
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" >
  </auditSources>
</configuration>
```

Any URL specified in the `auditSources` section must implement the [NuGet Server API V3 protocol](https://learn.microsoft.com/nuget/api/overview).
However, the server will only need to implement the [`VulnerabilityInfo` resource](https://learn.microsoft.com/en-us/nuget/api/vulnerability-info) (and service index), and can avoid implementing all the other resources that are required to be used as a package source.

NuGet will use same semantics for `auditSources` as it already has with `packageSources`.
This means that `auditSources` are accumulated from other nuget.config files, and `<clear />` can be used to remove them.
`auditSources` will also follow `packageSources` with regards to insecure http warnings, errors, and `allowInsecureConnections` configurations.
Sources that need credentials will use `packageSourceCredentials`, using the `key` as the lookup, just as `packageSources` does.
This means that credentials can be re-used by using the same key for a source that is defined in both `packageSources` and `auditSources`.

#### Restore summary

> When restore is run at normal or detailed verbosity, NuGet outputs detected nuget.config files and package sources (feeds) used.

Author: Should we filter out audit sources that do not provide a VulnerabilityInfo resource?
Or maybe flag them instead?

```text
Audit sources used:
    https://api.nuget.org/v3/index.json
    *https://packages.corp.contoso.com/nuget/index.json
    *C:\packages
*Audit source did not provide VulnerabilityInfo
```

Reviewer 2: Yes, only filter out sources that provide a proper resource. I think flagging could be acceptable too. Both convey the same information. I always err on showing the least amount of information until people are asking for explanations later.

It is very possible people may ask "why isn't packages.corp.contoso.com an audit source?" and we can document the reasons why or they can work with their package source provider to implement vulnerability info.

You've already documented the "server implementation" guide which will be helpful long term here.

Audit sources should be added.

```diff
 NuGet Config files used:
     C:\Users\zivkan\AppData\Roaming\NuGet\NuGet.Config

 Feeds used:
     https://api.nuget.org/v3/index.json
+

> +Audit sources used:

Reviewer: Is this shown when only package sources are used and auditing happens?

Author: "Audit sources used" will be shown when auditing happens, but regardless of whether package sources were used, or audit sources were defined in the nuget.config and therefore used.

Reviewer: Makes sense to me. Closing.

+    https://api.nuget.org/v3/index.json
```

> When `auditSources` do not have any configured sources, replay all of the package sources again under the audit sources used section, to make it explicit to customers that these sources are being used.

Author: I need to check the implementation, but if "feeds used" (`packageSources`) doesn't take into account package sources defind via msbuild, then the output will be incorrect. On the other hand, if only one project uses an additional source via MSBuild, then it might be confusing why the source is listed when most projects don't use it?  I'm really not a fan of NuGet settings via MSBuild, it's a shame it's adding additional complexity/uncertainty.

Also, another idea is have a message saying "Package sources were used as audit sources", and avoid duplicating the list again.

Anyone have feedback?

I'm actually learning towards the "Package sources were used as audit sources" message, I just don't feel like locally editing the markdown again and pushing another commit unless others agree.

What's better?

```text
Audit sources used:
    Package feed(s) used as audit source(s).
```

or just a one line

```text
Package feed(s) used as audit source(s).
```

Reviewer 2: Outside of Microsoft it is not uncommon for users to have 5+ sources. So multi-line is ideal here. I can share more survey results but I think I shared this specific question with you recently if you want to double check.

Reviewer 2: Also let's not use the word "feed". All of our stuff refers to "package sources".

Author: > So multi-line is ideal here.

When customers specify `auditSources` and there's more than 1, yes, this was absolutely my intention.

Are you saying that when customers don't specify any `auditSources`, and therefore NuGetAudit will use `packageSources`, we should duplicate the list, rather than state that the package sources list was re-used?

> Also let's not use the word "feed". All of our stuff refers to "package sources".

@JonDouglas are you implicitly suggesting that I change the existing "Feeds used" message to "Package sources used"?  I'm in favour of standardising on "package sources" and avoid the word "feed". But if we keep "Feeds useds" as a long standing existing message, I feel it makes more sense to use the same language that is being used just a few lines above.

Reviewer 2: in this context a package feed is a "package source". I am just calling out that we remain consistent to call these sources and not feeds.

Author: @JonDouglas sorry, I wasn't clear enough in my question.

The restore summary today already says "NuGet feeds used".  Do you want me to change this to "NuGet package sources used"?  This can be done completely independently of this spec, since it's an existing message.

Since we're planning to filter the "audit sources list" to only sources with a VDB, I've convinced myself that we should no longer have a 1 line "package sources used as audit sources". The intent was to reduce pointless duplication when the two lists are the same (1 empty line, 1 heading line, 1 line per source). But since this is a signal of which sources contain a VDB now, I no longer think the 1 liner is a good idea.

If we kept this "Package sources used a audit sources" message , then I think it's important to use the same wording as what was shown a few lines above. It would not be clear to beginners if it says "NuGet feeds used: ... Package sources used as audit sources".  But since I'm no longer proposing this message, it's not very relevant to this spec any longer.

Author: Given the summary text proposal is to filter out sources that don't have a VDB, I'm withdrawing this "Package sources used for audit sources" message idea. If you, or anyone else, believes the existing "NuGet feeds used" message should be changed to "NuGet package sources used", please create a new issue.

> If `NuGetAudit` is disabled for all projects in the solution, the audit sources used section of the summary should be omitted.

Author: This is something I only came up with as I was updating the spec. We didn't discuss this in the meeting. What do others think?

I'm not actually sure about the technical feasibility of checking all projects to see if NuGetAudit is enabled/disabled. But if a customer has disabled Audit for all their projects, it might look a bit weird if the restore summary still claims an audit source was used.

Should we have a message "NuGet Audit is disabled for all projects in the solution" instead?

Note that there's another related feature request:

* https://github.com/NuGet/Client.Engineering/issues/2340

I do not want this "audit sources used" mini-feature to become scope creep into implementing the entirety of all NuGetAudit related restore summary. But, this might be a quick win that can be changed when the "full" audit restore summary feature is designed & implemented.

Reviewer 2: Agreed. We should not show this experience unless users opt into the functionality. This higher level switch should influence the behavior down the line for other experiences.

We do not need a message here if the functionality is disabled. That's the user's choice and flexibility they desire (to not see those messages)

Also, customers are less likely to have unexpected changes if the app changes defaults.
On the other hand, when customers are happy to change with changing defaults, having an explicit config file makes it much more difficult for customers to learn about changed defaults, and from the package manager team's point of view, much more difficult to have customers adopt new defaults.

## Unresolved Questions

<!-- What parts of the proposal do you expect to resolve before this gets accepted? -->
<!-- What parts of the proposal need to be resolved before the proposal is stabilized? -->
<!-- What related issues would you consider out of scope for this proposal but can be addressed in the future? -->

All resolved during the spec review process.

## Future Possibilities

<!-- What future possibilities can you think of that this proposal would help with? -->

### Allow auditSources to point to a filesystem directory

In order to provide customers who are disconnected from nuget.org (and possibly the internet more widely), we could load the VDB from files on disk.
It would be up to customers to determine how to copy the VDB to their disconnected network/machine.

> We could consider a .NET tool, for example `nuget-audit-download` which knows how to communicate with nuget.org's V3 protocol, and merge results into a single file, so that it's easy for customers to obtain the file that needs to be copied.

Reviewer 6: We don't need the .NET tool for this section, right? It would just be like adding any folder feed, so nuget would utilize that feed to find the VDB?
You're just saying a secondary tool could be developed outside of nuget.exe/dotnet nuget?

Author: Correct. The only way today for customers to download NuGet's VDB is to understand our server API, then either develop a tool themselves, or open https://api.nuget.org/v3/index.json in their web browser, and follow the links to the VDB files and then save them.

What I'm implying with the last sentence is that if we think that customers who can't use nuget.org as an audit source is a common enough scenario for us to support officially, then we could provide a tool to make this as easy as possible for them. But there's no reason we have to. We could leave it for the community to develop themselves.

# NuGetAudit without nuget.org as a package source
<!-- Replace `Title` with an appropriate title for your design -->

- Author: [zivkan](https://github.com/zivkan)
- GitHub Issue: [12698](https://github.com/NuGet/Home/issues/12698)

## Summary

<!-- One-paragraph description of the proposal. -->

> This proposal adds a new `<auditSource>` to *NuGet.Config* files, allowing *NuGetAudit* to obtain known vulnerabilities without adding nuget.org as a package source.

Reviewer 7: How will this impact the browse/updates tab in PM UI (vulnerable adornments) and the `dotnet list --vulnerable` command?

Author: `dotnet list --vulnerable` uses package metadata (registration), not the vulnerability info resource. Therefore I consider it out of scope for this feature spec.

PM UI uses the vulnerability info resource, but only for installed, transitive packages. That code should be updated, so thanks for the reminder.

Reviewer 7: It might be worth mentioning the `dotnet list` limitation. If effectively means that `dotnet list --vulnerable` is stuck on `<packageSource>` because of a protocol difference. It might not be apparent to users that a VDB can't support `dotnet list`.

> This proposal adds a new `<auditSource>` to *NuGet.Config* files, allowing *NuGetAudit* to obtain known vulnerabilities without adding nuget.org as a package source.

Reviewer 7: Will we add a new sub-verb to `dotnet nuget add` like `dotnet nuget add audit-source`? We have `dotnet nuget add source` today which allows CLI or scripted addition of package sources. Wonder if we should start with CLI support for manipulating audit sources.

Author: This is already addressed at the end of the document, in the "tooling" header under the "future possibility" section.

Also see https://github.com/NuGet/Home/pull/12918#discussion_r1356721787

Reviewer 7: Oops, missed that. Thanks!

While nuget.org provides a VDB, there are several reasons why customers may not be using nuget.org as a package source.
This is especially true following the 2021 [blog post titled "Dependency confusion"](https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610) which significantly increased knowledge of multi-source package substitution supply chain risks.

While the design of NuGetAudit attempts to make it easy for upstreaming package sources to also upstream the `VulnerabilityInfo` resource, the NuGet team and customers are reliant of the feed implementor to implement the change.
An upstreaming package source is one that either aggregates multiple package sources or caches packages from another package source.
Upstreaming isn't a concept that exists in NuGet, so servers have to present to clients as if all data (packages, search results, vulnerability information) originates from itself.

However, there are also cases where customers are not using upstreaming feeds, but instead curate approved 3rd party packages and manually upload them to an internal feed.
There are also other scenarios where customers do not use nuget.org (directly) as a package source.

> It's possible for customers to use [Package Source Mapping](https://learn.microsoft.com/nuget/consume-packages/package-source-mapping) to add nuget.org as a package source, but configured in a way that NuGet will never restore packages from this package source.

Reviewer 7: The overlap of package source mapping and `<auditSource>` is interesting and raises some questions for me.

During the dependency confusion response, we encountered a bunch of cases where a common internal package name (internal to Microsoft or internal to 3rd party, both cases cropped up) overlapped with a NuGet.org package. With package source mapping, it allows the package consumer to say "no I don't want that NuGet.org thing, I want GenericLoggingLibrary from my own source". If the NuGet.org VDB starts reporting GenericLoggingLibrary as vulnerable, and the package consumer is using NuGet.org as an audit source, but they have a package source mapping saying GenericLoggingLibrary should come from their private feed, what should happen?

It seems like either the package source mapping would suppress the vulnerability report for that package since it's not the NuGet.org version (maybe) or it would show the warning to ere on the side of caution. It seems like the 2nd option is the safest to assume since ID conflicts between NuGet.org and remote are probably most often same content but it would be annoying for users that know the content is actually different. What would they do? Is there a good way to inform the users about this conflict and give them a suppression option, perhaps tailored to this case? It's not like they want to suppress a specific vulnerability, it's more like they want to say this package ID is not the same thing.

I guess this generalizes to the case when the user isn't using package source mapping at and only uses their internal feed, but an internal package ID just so happens to collide with a vulnerable range.

Reviewer 2: I think you are highlighting the importance of choice here.

I should have the choice to:

- Define where my packages come from (Source mapping)
- Define where my packages are audited from (VDB)
- Define the policy of how they will be audited (MSBuild prop/config)

I don't see any of these things as concerning, rather they are headed in the right directly. We would expect that any matching package ID + version from a reputable source implementing VDB would result in someone needing to look at that dependency, right?

Author: On one hand, NuGet has always been based around the assumption that packages are globally unique (and immutable). While package source mapping provides a step away from the globally unique assumption, NuGet will still share the same global packages folder for a particular package id+version combination, so if solution 1 gets  package from source 1, package source mapping will not prevent source 1's package that is already in the global packages folder when solution 2 expects the package to come from source 2. So, even with package source mapping, we're not fully away from the globally unique assumption.

Additionally, the same discussion (should package source mapping scope which packages are checked by NuGet Audit) is relevant to the existing NuGet Audit functionality, where VDBs are only downloaded from package sources, even if this feature spec is not implemented.
 
This is an interesting area, however, I consider it scope creep for this feature spec. Therefore, I'm going to resolve this PR conversation, but if someone thinks it's important enough, a new issue or feature spec can be created for that specific topic.

Like `<packageSources>`, `<auditSources>` can contain zero, one, or more sources, and supports `<clear />` to remove any `auditSources` inherited from parent *nuget.config* files.
When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.
Omitting `<clear />` will not use `packageSources` in addition to `auditSources`.

For example, consider a developer at Contoso using a company-internal source as a single package source, and using nuget.org as an audit source:

```xml
<configuration>
  <packageSources>
    <!-- Clear to ensure package sources not inherited from other config files -->
    <clear />
    <add key="contoso" value="https://internal.corp.contoso.com/packages/nuget/index.json" >
  </packageSources>
  <auditSources>
    <!-- Clear to ensure audit sources not inherited from other config files -->
    <clear />
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" >
  </auditSources>
</configuration>
```

Any URL specified in the `auditSources` section must implement the [NuGet Server API V3 protocol](https://learn.microsoft.com/nuget/api/overview).

> However, the server will only need to implement the [`VulnerabilityInfo` resource](https://learn.microsoft.com/en-us/nuget/api/vulnerability-info) (and service index), and can avoid implementing all the other resources that are required to be used as a package source.

Reviewer 7: I'm trying to understand the "need" here. If a feed does not implement the `VulnerabilityInfo` but it is set as an audit source, what happens? Will nuget error out? Warn? Silently act similar to a package source with no `VulnerabilityInfo`?

I'm thinking a user that is unaware of the protocol will naively put their ADO or other package source also as an audit source optimistically hoping for vulnerability info to start showing up. Can we guide them along the way to a successful config?

Reviewer 2: I'd err on silently doing nothing. But in the grand scheme, we should validate this and empower the user to rightful resolution.

Author: IMO, customers won't be able to "rely" on NuGetAudit, unless there's a way to warn or error when audit doesn't run (for example, no VDB could be obtained from any package source). If they expect audit to run as part of their pipeline, they should have a way to fail the pipeline if it doesn't run. And considering the only purpose of `auditSources` are to define sources used to get a VDB, I think it makes extra sense to warn if the configured audit source doesn't have a VDB.

However Joel's the first person who has given me feedback that appears to agree with my position, but since it's still only 1 person, so perhaps we can wait for more customer feedback.

Reviewer: I think it wouldn't be the worst thing even error when source configured as an "auditSource" doesn't have a VDB.

Reviewer 8: The same question occurred to me before I read Joel's question. 

Let's ignore for a moment that we may want to update the wordings of NU1900 and NU1905 with this feature or introduce new audit source-specific versions.

Currently, if I enable NuGetAudit, and no package source provides a VDB, then I expect [NU1905](https://learn.microsoft.com/nuget/concepts/auditing-packages#warning-codes).  The V3 `VulnerabilityResource` is an optional resource that each package source can implement (or not), and for backward compatibility reasons we don't require any package source to provide the resource.

The assumption for audit sources is that they do provide a VDB.  In fact, that is the _sole purpose_ of an audit source.

With this feature implemented, if I enable NuGetAudit and specify two audit sources, and one audit source provides a VDB but the other does not, I feel like that should still raise a warning.

Each audit source that does not provide a VDB should have its own warning.

Author: NU1905 was removed, for reasons I won't get into here, but it appears there is now some support for brining it back. Depending on the difficulty of implementation, I'll try to get it included in the implementation of this feature, or otherwise add a tracking issue for it.

> NuGet will use same semantics for `auditSources` as it already has with `packageSources`.

Reviewer 8: You established [here](https://github.com/NuGet/Home/pull/12918/files#diff-d4c82f8e89642e3f05039c1cb46e315bb78bf6ba400e7c6a6690e2fabc464413R61) that audit sources must implement the V3 protocol.

Package sources support a [`protocolVersion`](https://learn.microsoft.com/en-us/nuget/reference/nuget-config-file#packagesources) attribute.  You should be explicit about the behavior if an audit source has a `protocolVersion` attribute with a value other than 3 (e.g.:  2).  Do we ignore, warn, or error?

Reviewer 7: +1 on this point. I think this is a case where `<auditSources>` has difference semantics from `<packageSources>`. `protocolVersion` must be excluded or set to `3` and file-based sources should be rejected right?

IMO an error is best for this case since it's a new config value and we can therefore be strict from the beginning.

Author: If we're going to re-introduce NU1905 for when an audit source doesn't provide the VulnerabilityInfo resource, then it'll happen "automatically". The warning or error message might not be specifically calling out that v2 sources don't support it, but that's something we can document on the NU1905 docs page.

This means that `auditSources` are accumulated from other nuget.config files, and `<clear />` can be used to remove them.
`auditSources` will also follow `packageSources` with regards to insecure http warnings, errors, and `allowInsecureConnections` configurations.
Sources that need credentials will use `packageSourceCredentials`, using the `key` as the lookup, just as `packageSources` does.

> This means that credentials can be re-used by using the same key for a source that is defined in both `packageSources` and `auditSources`.

Reviewer 7: I'm also curious about the combinations of package sources and audit sources but maybe it's a little different from what Anand and Jon are asking about in [their comment thread](https://github.com/NuGet/Home/pull/12918/files#r1372141490). The combination that I think is most interesting to focus on is when there a no audit sources defined and 1 or more package sources with no VDB. When a user with this config updates to the client version that carries this change, how will they be guided to adding NuGet.org as the audit source (which seems like a very good suggestion for someone to give them)? Or is there a default where we inject NuGet.org as `<auditSources>` perhaps at the user level allowing them to override that with a audit sources `<clear />` at the repo level?

This is maybe along the same lines of [Svetlana's comment](https://github.com/NuGet/Home/pull/12918/files#r1376881089) but I thought she was more asking about how new NuGet.config files get initialized and less about adoption of this feature by existing nuget.configs.

Reviewer 2: IMO the logic should be.

1. If you have a package source with a VDB, it is implicitly added to the auditsources (not actually in the nuget.config though)
2. If you don't have a package source with a VDB, you explicitly add it to the auditsources in the nuget.config.
3. If you have multiple auditsources defined, maybe there is a question of whether they audit their own sources accordingly or if one should gesture what source to audit for all packages? I think we are seeing this from the central registry perspective today but in the far future this could be used in different contexts. 

Apologies as I'm catching up.

Author: Jon, I interpret your comment as being orthogonal to Joel's question.

> The combination that I think is most interesting to focus on is when there a no audit sources defined and 1 or more package sources with no VDB.

I think having at least one source that doesn't provide a VDB is going to be very common, and would be too spammy to have any output for. Also probably wouldn't be productive if a different source does provide a VDB.

However, a restore where none of the sources provided a VDB is something I think can be reasonable to provide feedback on. Currently, I'm thinking maybe the end of restore summary could have a message something like:

```text
Audit sources used:
   None, see https://aka.ms/nuget/audit#sources for more information
```

The part of the spec that proposes the restore summary change says that if audit isn't run, it shouldn't output this part of the message. However, maybe we should only omit this message if Audit is disabled, rather than no audit sources used.

However, an issue with this is that this message is only output at normal verbosity, and both VS and the `dotnet` CLI default to minimal verbosity, so many customers won't see the message.

> Or is there a default where we inject NuGet.org as <auditSources> perhaps at the user level allowing them to override that with a audit sources <clear /> at the repo level?

That's something I proposed as an idea earlier, but other people in the team didn't like it. I still think it's a good idea, and in line with what all other package manager ecosystems are doing. 🤷 Frankly, I think it's a better for package sources as well, compared to "create default config if it doesn't already exist", and would have avoided all the issues with had with the config tracking file and auto-adding nuget.org in the past.

Author: I still think that having default sources not require any nuget.config declarations is a good idea. In fact, I feel strongly enough that I created a feature request/DCR issue:

* https://github.com/NuGet/Home/issues/13027

#### Restore summary

When restore is run at normal or detailed verbosity, NuGet outputs detected nuget.config files and package sources (feeds) used.
Audit sources should be added.

```diff
 NuGet Config files used:
     C:\Users\zivkan\AppData\Roaming\NuGet\NuGet.Config

 Feeds used:
     https://internal.corp.contoso.com/packages/nuget/index.json
+
+Audit sources used:
+    https://api.nuget.org/v3/index.json
```

The "Audit sources used" section should ideally list sources where a VDB was successfully obtained.
This depends on the feasibility of implementation.
For example, if two package sources are defined, and no audit sources are defined by the *nuget.config* file, but only one of the two sources has a VDB, then the restore summary will only list the one source with a VDB under "Audit sources used".

If `NuGetAudit` is disabled for all projects in the solution, the audit sources used section of the summary should be omitted.

> ### Technical explanation

Reviewer 7: Will the HTTP cache for the audit source be shared with the package source of the same URL?

Author: yes, as written in the functional explanation section, audit sources use all the same semantics as package sources.

> ### Technical explanation

Reviewer 7: Will proxy settings apply to the audit source in the same way as package source?

Author: yes, as written in the functional explanation section, audit sources use all the same semantics as package sources.

Reviewer 7: Sounds good, I wasn't sure how far "same semantics" went in this case 🙂

<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->

Not using `packageSources` for VDB when there's at least one source in `auditSources` is a performance optimization.

> Since we expect customers to use `auditSources` when their `packageSources` don't already provide a VDB, having restore spend time making HTTP requests to the package sources is unlikely to provide any benefit.

Reviewer 7: How will a user know their package source doesn't provide a VDB?

Author: It'll be indirect, but assuming the end of restore summary isn't infeasible to implement, they can check that the source URL is listed under "Audit sources used"

Reviewer 7: This seems reasonable for package source since it would be kind of breaking to have a warning/error if audit is enabled (which I recall it is by default) and they have no VDB. Thanks for clarifying!

While downloading the VDB from audit sources is likely to take more time than verifying that package sources don't have a vulnerability resource, if any package source is slow, downloading the VDB from the audit source might complete first.
Additionally, in the case where NuGet doesn't otherwise need to communicate with any package source, checking the package sources for a VDB might trigger a credential provider.

## Drawbacks

<!-- Why should we not do this? -->

> Customers who appear functionally network isolated from nuget.org (for example, by blocking nuget.org at the firewall) will still not be able to use this feature.

Reviewer 8: After reading this and [this](https://learn.microsoft.com/dotnet/core/tools/dotnet-restore#audit-for-security-vulnerabilities), it's my (incorrect?) understanding that NuGetAudit currently requires NuGet.org as a package source, that even if I use my own package source that provides the V3 `VulnerabilityResource` NuGetAudit won't work.  Is that true?

I suspect no, it isn't true, that any package source that implements the V3 protocol sufficiently (including the `VulnerabilityResource` resource) could serve as a VDB source (as a package source) in .NET 8 SDK today.  This would be consistent with the V3 design:  any V3 package source can pick and choose which resources they want to support.  The docs (linked above) don't do a good job explaining that NuGet.org is recommended because it's the only package source we know that implements the `VulnerabilityResource` resource, and that any package source that implements the resource could be used in addition to or instead of NuGet.org.

```suggestion
Customers who appear functionally network isolated from nuget.org (for example, by blocking nuget.org at the firewall) will only be able to use this feature if their `auditSources` are accessible.
```

Author: > Is that true?

No, as you suspect, Audit works with any source that supports the vulnerability info resource. I guess the person who added that info to the dotnet docs misunderstood or was provided incorrect information. We'll get the docs updated:

* https://github.com/dotnet/docs/issues/39212

Pypi's `pip` has [an audit command](https://pypi.org/project/pip-audit/), and `--index-url` and `--extra-index-url` appear to enable using custom audit sources.

Rust's cargo audit command has [configuration for the advisory DB URL](https://docs.rs/cargo-audit/latest/cargo_audit/config/struct.DatabaseConfig.html)

Note that for these other package ecosystems, they do not appear to need any explicit configuration file in order to specify default sources (either package source, or audit source).
Instead, defaults appear hard-coded into the app, and it requires explicit configuration to remove.

Contrast this to NuGet's design where NuGet creates a default configuration file at first use, and then requires sources to be explicitly defined.
However, NuGet settings that fall under the `<config>` section of the *nuget.config* file do have hard-coded defaults that do not need to be explicitly defined in any config file.

> Both choices have advantages and disadvantages.

Reviewer 9: That's a great discussion point, @zivkan . When a new nuget.config is created, should it be created with nuget.org as the default audit source? IMHO, it should, so that in case a user removes nuget.org from the packageSources list, they will continue getting the auditing functionality.

Author: IMO no, it should not, unless the proposal to use packageSources only as a fallback when auditSources is not defined is dropped. See this conversation: https://github.com/NuGet/Home/pull/12918#discussion_r1372141490

If we change this spec so that `auditSources` is always additive to `packageSources`, then it makes sense.

If we change this spec so that `packageSources` is never used for NuGetAudit (my preference, but is a breaking change), then it also makes sense. Although we'd also want to think about how customers who have existing nuget.config files will adopt the feature, in which case I'd say we'd be better off to hardcode nuget.org as an implicit auditSource that needs `<clear/>` to remove.

However, with the current design proposal, adding nuget.org to the default nuget.config's `auditSources` will give a false impression that NuGet is non-deterministic, depending if the customer is using a machine with an existing user-profile nuget.org or started from a clean machine without one.

Author: Due to unpopular feedback, package sources will no longer be a fallback for audit. So now this suggestion makes more sense.

However, I still think it's not the best solution for the reasons I listed in:

* https://github.com/NuGet/Home/issues/13027

I strongly feel that NuGet should "hard code" defaults in NuGet.Configuration, and customers should require a `<auditSources><clear /></auditSources>` to remove nuget.org as an audit source. This will prevent PowerShell, or any app that is not a NuGet team official tool (including other Microsoft products), from creating default nuget.config files that prevent nuget.org from being added as a default audit source.
